### PR TITLE
add missing yargs dependency on filesystem

### PR DESCRIPTION
**Pull request by amiramw**
Signed-off-by: Amiram Wingarten <amiram.wingarten@sap.com>

#### What it does

Add missing yargs dependency on filesystem.
Code use is here: https://github.com/eclipse-theia/theia/blob/9d2dcd0a0923d7f7d8def0b7377fc9667706c967/packages/filesystem/src/node/nsfw-watcher/index.ts#L17

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by kittaakos** [comment]
> Add missing yargs dependency on filesystem.
> Code use is here:

I do not like transitive dependencies either, but why this change necessary? `yargs` and its typing is available from `@theia/core`:

https://github.com/eclipse-theia/theia/blob/101c702ee143b26167a7b0b81ac38a52f5dd72bf/packages/core/package.json#L24

https://github.com/eclipse-theia/theia/blob/101c702ee143b26167a7b0b81ac38a52f5dd72bf/packages/core/package.json#L55

What does this PR solve? I cannot imagine a Theia app without the `core`. Thanks!

**Comment by amiramw** [comment]
@kittaakos @vince-fugnitto 

I have a dockerfile that builds theia with our extensions in production mode and for some reason it recently doesn't hoist yargs to upper node_modules. It is only created under node_modules/@theia/core/node_modules.

I suspect that it is related to recent changes in my project that introduced other yargs dependencies with different versions.

https://github.com/eclipse-theia/theia/pull/5832 has some discussion about the whitelisting. BTW, we have the same issue with fs-extra and we are forced to explicitly declare its dependency in our root package.json.

**Comment by kittaakos** [comment]
Thanks for the details, @amiramw 👍 

> and for some reason

We should find the reason. Did you check with `yarn why yargs`?

> theia with our extensions

These things have to be handled with `resolutions` in your `package.json` and not in Theia.

This transitive dependency question comes back all the time, I do not like them but the majority of the committers decided to not use them, so I tend to reject this change. Otherwise, we open Pandora's box.

> we have the same issue with fs-extra

I can confirm this.

> and we are forced to explicitly declare its dependency in our root package.json.

You have to use [`resolutions`](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/) and not redeclare the dependency. See an example [here](https://github.com/theia-ide/theia-apps/blob/a54aaa676e3db07d22ab75dde9f3447576135b4d/theia-electron/package.json#L79-L81) or an electron-specific explanation [here](https://spectrum.chat/theia/general/our-theia-electron-builder-app-no-longer-starts~f5cf09a0-6d88-448b-8818-24ad0ec2ee7c).

**Comment by amiramw** [comment]
> This transitive dependency question comes back all the time, I do not like them but the majority of the committers decided to not use them, so I tend to reject this change. Otherwise, we open Pandora's box.

This sounds fragile. Some of those dependencies are pretty popular and downstream projects bump into this issue. I'll try the resolution solution. It means that  if I want to be sure that theia works with the correct versions then I should add resolution to each package, no? Otherwise I may run with different major version of some dependencies which no one intended to run with.

Also the way to find those dependencies is to go over theia code.

Have you considered build time solution for that? It should be pretty easy to force that all packages use the same version of a given dependency for example.

**Comment by kittaakos** [comment]
> This sounds fragile

I agree. But re-declaring a problematic dependency at a single location is not a fix. We need a proper solution or you have to use `resolutions`.

> add resolution to each package

`resolutions` works only in the root `package.json`. Otherwise, it has no effect.

> Have you considered build time solution for that?

Something like [this](https://github.com/eclipse-theia/theia/pull/7303)?

**Comment by amiramw** [comment]
Adding `"**/yargs": "^11.1.0"` as `resolution` didn't work for me. ts copmilation failed to find some modules.

```
$ yarn why yargs
yarn why v1.22.4
[1/4] Why do we have the module "yargs"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "yargs@11.1.1"
info Has been hoisted to "yargs"
info Reasons this module exists
   - "workspace-aggregator-9b16c7ea-b365-4f3a-a41d-f5265143c60e" depends on it
   - Hoisted from "_project_#browser-app#@theia#cli#yargs"
   - Hoisted from "_project_#@sap#theia-hosted-plugin-manager#@theia#core#yargs"
   - Hoisted from "_project_#browser-app#@theia#cli#puppeteer-to-istanbul#yargs"
   - Hoisted from "_project_#browser-app#@theia#cli#@theia#application-manager#webpack-cli#yargs"
info Disk size without dependencies: "480KB"
info Disk size with unique dependencies: "848KB"
info Disk size with transitive dependencies: "1.7MB"
info Number of shared dependencies: 38
=> Found "@lerna/cli#yargs@12.0.5"
info This module exists because "_project_#lerna#@lerna#cli" depends on it.
info Disk size without dependencies: "304KB"
info Disk size with unique dependencies: "672KB"
info Disk size with transitive dependencies: "1.52MB"
info Number of shared dependencies: 38
=> Found "jest-cli#yargs@15.4.1"
info This module exists because "_project_#jest#jest-cli" depends on it.
info Disk size without dependencies: "376KB"
info Disk size with unique dependencies: "728KB"
info Disk size with transitive dependencies: "984KB"
info Number of shared dependencies: 21
=> Found "mocha#yargs@13.3.2"
info This module exists because "_project_#browser-app#@theia#cli#mocha" depends on it.
info Disk size without dependencies: "320KB"
info Disk size with unique dependencies: "656KB"
info Disk size with transitive dependencies: "928KB"
info Number of shared dependencies: 21
=> Found "lockfile-lint#yargs@13.3.2"
info This module exists because "_project_#@sap#theia-wing-plugin-ext#verdaccio#lockfile-lint" depends on it.
info Disk size without dependencies: "320KB"
info Disk size with unique dependencies: "656KB"
info Disk size with transitive dependencies: "928KB"
info Number of shared dependencies: 21
=> Found "jest-runtime#yargs@15.4.1"
info This module exists because "_project_#jest#@jest#core#jest-runtime" depends on it.
info Disk size without dependencies: "376KB"
info Disk size with unique dependencies: "728KB"
info Disk size with transitive dependencies: "984KB"
info Number of shared dependencies: 21
=> Found "yargs-unparser#yargs@13.3.2"
info This module exists because "_project_#browser-app#@theia#cli#mocha#yargs-unparser" depends on it.
info Disk size without dependencies: "320KB"
info Disk size with unique dependencies: "656KB"
info Disk size with transitive dependencies: "928KB"
info Number of shared dependencies: 21
=> Found "electron-rebuild#yargs@14.2.3"
info This module exists because "_project_#browser-app#@theia#cli#@theia#application-manager#electron-rebuild" depends on it.
info Disk size without dependencies: "320KB"
info Disk size with unique dependencies: "672KB"
info Disk size with transitive dependencies: "928KB"
info Number of shared dependencies: 21
=> Found "glob-all#yargs@15.4.1"
info This module exists because "_project_#browser-app#@theia#cli#@theia#application-manager#webpack-cli#glob-all" depends on it.
info Disk size without dependencies: "376KB"
info Disk size with unique dependencies: "728KB"
info Disk size with transitive dependencies: "984KB"
info Number of shared dependencies: 21
Done in 1.96s.
```

I'm pretty sure that upgrading jest caused this regression as it now depends on yargs ^15.3.1 and probably forcing older version breaks the build.

Any alternative to adding yargs as dependency?

**Comment by kittaakos** [comment]
> Adding `"**/yargs": "^11.1.0"` as `resolution` didn't work for me

Why is that? The `yarn why` log you have attached shows `yargs@11.1.0` was hoisted:

> => Found "yargs@11.1.1"
> info Has been hoisted to "yargs"
> info Reasons this module exists
>    - "workspace-aggregator-9b16c7ea-b365-4f3a-a41d-f5265143c60e" depends on it
>    - Hoisted from "_project_#browser-app#@theia#cli#yargs"
>    - Hoisted from "_project_#@sap#theia-hosted-plugin-manager#@theia#core#yargs"
>    - Hoisted from "_project_#browser-app#@theia#cli#puppeteer-to-istanbul#yargs"
>    - Hoisted from "_project_#browser-app#@theia#cli#@theia#application-manager#webpack-cli#yargs"


> . ts copmilation failed to find some modules.

It's very cryptic: hard to say anything without seeing the error. I wonder what would be the difference if we redeclare `yargs` in `@theia/filesystem`.

> I'm pretty sure that upgrading jest caused this regression

I think it's fair to say Theia is not responsible for taking care of your `jest` dependencies:
```
theia % git rev-parse --short HEAD && yarn why jest
9d2dcd0a0
yarn why v1.22.4
[1/4] 🤔  Why do we have the module "jest"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
error We couldn't find a match!
✨  Done in 0.94s.
```

To sum up:
 - You can explain to us how does the redeclaration of the `yargs` dependency solve this issue. Maybe your right; I just do not understand it yet. Please explain.
 - You can try to bump up `yargs` in Theia and see if it solves your problem.
 - You can also try to find someone who will approve this change as-is; I am not going to object.

File: packages/filesystem/package.json [hunk 1]
```diff
@@ -60,7 +61,8 @@
     "test:watch": "theiaext test:watch"
   },
   "devDependencies": {
-    "@theia/ext-scripts": "^1.4.0"
+    "@theia/ext-scripts": "^1.4.0",
+    "@types/yargs": "^11.1.0"
```
Comment (kittaakos) [inline packages/filesystem/package.json:65]: This should go to `dependencies`.

Comment (amiramw) [reply packages/filesystem/package.json:65]: why @types should be in dependencies? aren't they required only on dev?

Comment (kittaakos) [reply packages/filesystem/package.json:65]: By convention, we put the typings next to the dependency. See: https://github.com/eclipse-theia/theia/commit/81fb0c61ddbfb6274abd1918aad81bb79eb8c4a9

Comment (akosyakov) [reply packages/filesystem/package.json:65]: We add types to dependencies that downstream don't need to add them as dev dependencies. Yes It is a convention.

**Comment by akosyakov** [comment]
@amiramw You should be able to use resolutions  in root package.json like:
```json
"resolutions": {
    "@theia/**/yargs": "^11.1.0"
}
```

I don't really know the good solution here:
- if we only keep `yargs` in `@theia/core` then there can be a version mismatch between hoisted and core version
- if we recommend to redeclare dependencies then there can be a version mismatch between downstream and core version

In both cases work is required to find an eliminate the mismatch, either with resolutions or redeclaring another versions in downstream dependencies.

I don't mind accepting this PR though. If downstream Theia extensions don't declare own version of yargs, which I assume they don't do, it should not harm.

**Comment by marechal-p** [comment]
Overall I really dislike relying on hoisting. It is fragile and while there are workarounds, they are workarounds.

@akosyakov since we expect `@theia/core` to provide basic dependencies that extensions can rely upon, why not having something re-export the dependencies from the package itself, instead of relying on hoisting?

Something like:

```ts
// some-index.ts

export import * as depA from 'depA';
export import * as depB from 'depB';
// ...
```

This way someone that depends on `@theia/core` would need to grab the common dependencies from this specific file that re-exports everything useful to it. I think it would be more robust than what we have now, wdyt?

**Comment by akosyakov** [comment]
@marechal-p yes, we could try to play with it.

**Comment by kittaakos** [comment]
I propose closing this PR. Unfortunately, we have to use `resolutions` in some cases; this is how it is. Any objections, @amiramw?

**Comment by amiramw** [comment]
No. Closing.

**Pull request closed without merging**